# **Breast Cancer Detection Using Neural Networks**

This notebook implements a neural network using PyTorch to predict whether a breast tumor is **malignant** or **benign** based on clinical features from the sklearn breast cancer dataset.

The purpose of this project is to demonstrate how neural networks can assist in early detection of breast cancer using patient data. It also serves as a foundation for exploring AI applications in healthcare and experimenting with different network architectures.


**Importing the Dependencies**

In [216]:
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

**Device Configuration**

In [217]:
# check for CUDA availability
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Using device: {device }")

Using device: cuda


**Data collection and Preprocessing**

In [218]:
# load the breast cancer dataset
data = load_breast_cancer()
x = data.data
y = data.target

In [219]:
print(y[:5])

[0 0 0 0 0]


**Train Test Split**

In [220]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [221]:
print(x.shape)
print(x_train.shape)
print(x_test.shape)

(569, 30)
(455, 30)
(114, 30)


In [222]:
# Standardize the data using standard scaler
scaler = StandardScaler()

x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

In [223]:
type(x_train)

numpy.ndarray

In [224]:
# convert data to PyTorch Tensor and move it to GPU
x_train = torch.tensor(x_train, dtype=torch.float32).to(device)
y_train = torch.tensor(y_train, dtype=torch.float32).to(device)
x_test = torch.tensor(x_test, dtype=torch.float32).to(device)
y_test = torch.tensor(y_test, dtype=torch.float32).to(device)

**Neural Network Architecture**

In [225]:
# device the neural network architecture

class NeuralNet(nn.Module):

  def __init__(self, input_size, hidden_size, output_size):
    super(NeuralNet, self).__init__()
    self.fcl = nn.Linear(input_size, hidden_size)
    self.relu = nn.ReLU()
    self.fc2 = nn.Linear(hidden_size, output_size)
    self.sigmoid = nn.Sigmoid()

  def forward(self , x):
    out = self.fcl(x)
    out = self.relu(out)
    out = self.fc2(out)
    out = self.sigmoid(out)
    return out

In [226]:
# define Hyperparameters
input_size = x_train.shape[1]
hidden_size = 64
output_size = 1
learning_rate = 0.001
num_epochs = 100

In [227]:
# initialize the neural network and move it the GPU
model = NeuralNet(input_size, hidden_size , output_size).to(device)

In [228]:
# define loss and the optimizer
criterion = nn.BCELoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

**Training the Neural Network**

In [233]:
# training the model
for epoch in range(num_epochs):
  model.train()
  optimizer.zero_grad()
  outputs = model(x_train)
  loss = criterion(outputs, y_train.view(-1,1))
  loss.backward()
  optimizer.step()

  #calculate accurancy
  with torch.no_grad():
    predicted = outputs.round()
    correct = (predicted == y_train.view(-1,1)).float().sum()
    accuracy = correct/y_train.size(0)

    if (epoch + 1) % 10 == 0:
      print(f"Epoch [{epoch+1}/{num_epochs}], "
            f"Loss: {loss.item():.4f}, "
            f"Accuracy: {accuracy.item() * 100:.2f}%")

Epoch [10/100], Loss: 0.1071, Accuracy: 97.14%
Epoch [20/100], Loss: 0.0993, Accuracy: 97.36%
Epoch [30/100], Loss: 0.0927, Accuracy: 97.80%
Epoch [40/100], Loss: 0.0870, Accuracy: 97.80%
Epoch [50/100], Loss: 0.0821, Accuracy: 98.02%
Epoch [60/100], Loss: 0.0778, Accuracy: 98.24%
Epoch [70/100], Loss: 0.0739, Accuracy: 98.46%
Epoch [80/100], Loss: 0.0705, Accuracy: 98.46%
Epoch [90/100], Loss: 0.0674, Accuracy: 98.46%
Epoch [100/100], Loss: 0.0647, Accuracy: 98.46%


**Model Evaluation**

Evaluation on Training Data

In [234]:
model.eval()
with torch.no_grad():
  Outputs = model(x_train)
  predicted - outputs.round()
  correct = (predicted == y_train.view(-1,1)).float().sum()
  accuracy = correct/y_train.size(0)
  print(f"Accuracy on training data: {accuracy.item() * 100:.2f}%")

Accuracy on training data: 98.46%


Evaluation on Testing Data

In [231]:
model.eval()
with torch.no_grad():
  outputs = model(x_test)
  predicted = outputs.round()
  correct = (predicted == y_test.view(-1,1)).float().sum()
  accuracy = correct/y_test.size(0)
  print(f"Accuracy on testing data: {accuracy.item() * 100:.2f}%")

Accuracy on testing data: 98.25%


**## Summary & Model Performance**

**Breast Cancer Detection Using Neural Networks**

This project used a feedforward neural network (ANN) with one hidden layer to classify breast tumors. The model shows that neural networks can achieve high accuracy using clinical features and can be applied for medical AI tasks.

**Model Performance:**

- **Accuracy on Training Data:** 98.46%  
- **Accuracy on Testing Data:** 98.25%
